File: src/test/java/toby/command/commands/moderation/KickCommandTest.java
package toby.command.commands.moderation;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Mentions;
import net.dv8tion.jda.api.entities.channel.unions.AudioChannelUnion;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.requests.restaction.AuditableRestAction;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import toby.command.CommandContext;
import toby.command.CommandTest;

import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class KickCommandTest implements CommandTest {

    KickCommand KickCommand;
    
    @BeforeEach
    void setUp() {
        setUpCommonMocks();
        KickCommand = new KickCommand();
    }

    @AfterEach
    void tearDown() {
        tearDownCommonMocks();
    }

    @Test
    void test_KickWithValidPermissions_kicksEveryoneInChannel() {
        //Arrange
        CommandContext commandContext = new CommandContext(event);
        kickSetup(true, true, List.of(targetMember));

        //Act
        KickCommand.handle(commandContext, requestingUserDto, 0);

        //Assert
        verify(event, times(1)).getGuild();
        verify(guild, times(1)).kick(targetMember);

    }

    @Test
    void test_KickWithValidPermissionsAndMultipleMembers_kicksEveryoneInChannel() {
        //Arrange
        CommandContext commandContext = new CommandContext(event);
        kickSetup(true, true, List.of(member, targetMember));

        //Act
        KickCommand.handle(commandContext, requestingUserDto, 0);

        //Assert
        verify(event, times(1)).getGuild();
        verify(guild, times(1)).kick(member);
        verify(guild, times(1)).kick(targetMember);

    }


    @Test
    void test_KickWithInvalidBotPermissions_throwsError() {
        //Arrange
        CommandContext commandContext = new CommandContext(event);
        kickSetup(false, true, List.of(targetMember));

        //Act
        KickCommand.handle(commandContext, requestingUserDto, 0);

        //Assert
        verify(event, times(1)).getGuild();
        verify(interactionHook, times(1)).sendMessageFormat(eq("I'm not allowed to kick %s"), eq(targetMember));

    }

    @Test
    void test_KickWithInvalidUserPermissions_throwsError() {
        //Arrange
        CommandContext commandContext = new CommandContext(event);
        kickSetup(true, false, List.of(targetMember));

        //Act
        KickCommand.handle(commandContext, requestingUserDto, 0);

        //Assert
        verify(event, times(1)).getGuild();
        verify(interactionHook, times(1)).sendMessageFormat(eq("You can't kick %s"), eq(targetMember));

    }

    private static void kickSetup(boolean botKickOthers, boolean memberKickOthers, List<Member> mentionedMembers) {
        AudioChannelUnion audioChannelUnion = mock(AudioChannelUnion.class);
        GuildVoiceState guildVoiceState = mock(GuildVoiceState.class);
        AuditableRestAction auditableRestAction = mock(AuditableRestAction.class);
        OptionMapping optionMapping = mock(OptionMapping.class);
        Mentions mentions = mock(Mentions.class);
        when(event.getOption("users")).thenReturn(optionMapping);
        when(optionMapping.getMentions()).thenReturn(mentions);
        when(mentions.getMembers()).thenReturn(mentionedMembers);
        when(member.canInteract(any(Member.class))).thenReturn(true);
        when(botMember.hasPermission(Permission.KICK_MEMBERS)).thenReturn(botKickOthers);
        when(botMember.canInteract(any(Member.class))).thenReturn(botKickOthers);
        when(member.getVoiceState()).thenReturn(guildVoiceState);
        when(member.hasPermission(Permission.KICK_MEMBERS)).thenReturn(memberKickOthers);
        when(guildVoiceState.getChannel()).thenReturn(audioChannelUnion);
        when(guild.kick(any())).thenReturn(auditableRestAction);
        when(auditableRestAction.reason(eq("because you told me to."))).thenReturn(auditableRestAction);
    }
}

File: src/test/kotlin/toby/command/commands/moderation/KickCommandTest.kt
package toby.command.commands.moderation

import io.mockk.*
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Mentions
import net.dv8tion.jda.api.entities.channel.unions.AudioChannelUnion
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import net.dv8tion.jda.api.requests.restaction.AuditableRestAction
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import toby.command.CommandContext
import toby.command.CommandTest
import toby.command.CommandTest.Companion.botMember
import toby.command.CommandTest.Companion.event
import toby.command.CommandTest.Companion.guild
import toby.command.CommandTest.Companion.member
import toby.command.CommandTest.Companion.requestingUserDto
import toby.command.CommandTest.Companion.targetMember

internal class KickCommandTest : CommandTest {
    private lateinit var kickCommand: KickCommand

    @BeforeEach
    fun setUp() {
        setUpCommonMocks()
        kickCommand = KickCommand()
    }

    @AfterEach
    fun tearDown() {
        tearDownCommonMocks()
        unmockkAll()
    }

    @Test
    fun test_KickWithValidPermissions_kicksEveryoneInChannel() {
        // Arrange
        val commandContext = CommandContext(event)
        kickSetup(botKickOthers = true, memberKickOthers = true, mentionedMembers = listOf(targetMember))

        // Act
        kickCommand.handle(commandContext, requestingUserDto, 0)

        // Assert
        verify(exactly = 1) { event.guild }
        verify(exactly = 1) { guild.kick(targetMember) }
    }

    @Test
    fun test_KickWithValidPermissionsAndMultipleMembers_kicksEveryoneInChannel() {
        // Arrange
        val commandContext = CommandContext(event)
        kickSetup(
            botKickOthers = true,
            memberKickOthers = true,
            mentionedMembers = listOf(member, targetMember)
        )

        // Act
        kickCommand.handle(commandContext, requestingUserDto, 0)

        // Assert
        verify(exactly = 1) { event.guild }
        verify(exactly = 1) { guild.kick(member) }
        verify(exactly = 1) { guild.kick(targetMember) }
    }

    @Test
    fun test_KickWithInvalidBotPermissions_throwsError() {
        // Arrange
        val commandContext = CommandContext(event)
        kickSetup(botKickOthers = false, memberKickOthers = true, mentionedMembers = listOf(targetMember))

        // Act
        kickCommand.handle(commandContext, requestingUserDto, 0)

        // Assert
        verify(exactly = 1) { event.guild }
        verify(exactly = 1) { event.hook.sendMessage(any<String>()) }
    }

    @Test
    fun test_KickWithInvalidUserPermissions_throwsError() {
        // Arrange
        val commandContext = CommandContext(event)
        kickSetup(botKickOthers = true, memberKickOthers = false, mentionedMembers = listOf(targetMember))

        // Act
        kickCommand.handle(commandContext, requestingUserDto, 0)

        // Assert
        verify(exactly = 1) { event.guild }
        verify(exactly = 1) { event.hook.sendMessage(any<String>()) }
    }

    companion object {
        private fun kickSetup(botKickOthers: Boolean, memberKickOthers: Boolean, mentionedMembers: List<Member>) {
            val audioChannelUnion = mockk<AudioChannelUnion>()
            val guildVoiceState = mockk<GuildVoiceState>()
            val auditableRestAction = mockk<AuditableRestAction<Void>>()
            val optionMapping = mockk<OptionMapping>()
            val mentions = mockk<Mentions>()

            every { event.getOption("users") } returns optionMapping
            every { optionMapping.mentions } returns mentions
            every { mentions.members } returns mentionedMembers
            every { member.canInteract(any<Member>()) } returns true
            every { botMember.hasPermission(Permission.KICK_MEMBERS) } returns botKickOthers
            every { botMember.canInteract(any<Member>()) } returns botKickOthers
            every { member.voiceState } returns guildVoiceState
            every { member.hasPermission(Permission.KICK_MEMBERS) } returns memberKickOthers
            every { guildVoiceState.channel } returns audioChannelUnion
            every { guild.kick(any<Member>()) } returns auditableRestAction
            every { auditableRestAction.reason("because you told me to.") } returns auditableRestAction
            every { auditableRestAction.queue(any(), any()) } just Runs
        }
    }
}
